File: modules/agent/src/serializers/task.py
"""Serializers for task-related models."""

from typing import Optional, List


def format_datetime(value) -> Optional[str]:
    """Format datetime to ISO string."""
    if not value:
        return None
    return value.isoformat() if hasattr(value, "isoformat") else str(value)


def format_date(value) -> Optional[str]:
    """Format date to ISO string."""
    if not value:
        return None
    date_str = value.isoformat() if hasattr(value, "isoformat") else str(value)
    return date_str[:10] if date_str else None


def format_decimal(value) -> Optional[float]:
    """Format Decimal to float."""
    if value is None:
        return None
    return float(value)


def to_paged_response_with_scope(count: int, page: int, limit: int, items: List, scope: Optional[dict] = None) -> dict:
    """Convert to paged response format with optional scope."""
    response = {
        "items": items,
        "currentPage": page,
        "totalPages": max(1, (count + limit - 1) // limit),
        "total": count,
        "pageSize": limit,
    }
    if scope:
        response["scope"] = scope
    return response


def task_to_list_response(task, entity_info: Optional[dict] = None) -> dict:
    """Convert Task model to dictionary - optimized for list/table view."""
    if entity_info is None:
        entity_info = {
            "type": None,
            "id": None,
            "display_name": None,
            "url": None,
            "lead_type": None,
        }

    return {
        "id": task.id,
        "title": task.title,
        "due_date": format_date(task.due_date),
        "status": {
            "id": task.current_status.id,
            "name": task.current_status.name,
        } if task.current_status else None,
        "priority": {
            "id": task.priority.id,
            "name": task.priority.name,
        } if task.priority else None,
        "entity": entity_info,
        "created_at": format_datetime(task.created_at),
        "updated_at": format_datetime(task.updated_at),
    }


def task_to_detail_response(task, entity_info: Optional[dict] = None) -> dict:
    """Convert Task model to dictionary - full detail view."""
    if entity_info is None:
        entity_info = {
            "type": None,
            "id": None,
            "display_name": None,
            "url": None,
            "lead_type": None,
        }

    return {
        "id": task.id,
        "title": task.title,
        "description": task.description,
        "start_date": format_date(task.start_date),
        "due_date": format_date(task.due_date),
        "estimated_hours": format_decimal(task.estimated_hours),
        "actual_hours": format_decimal(task.actual_hours),
        "complexity": task.complexity,
        "sort_order": task.sort_order,
        "completed_at": format_datetime(task.completed_at),
        "assigned_to_individual_id": task.assigned_to_individual_id,
        "status": {
            "id": task.current_status.id,
            "name": task.current_status.name,
        } if task.current_status else None,
        "priority": {
            "id": task.priority.id,
            "name": task.priority.name,
        } if task.priority else None,
        "entity": entity_info,
        "created_at": format_datetime(task.created_at),
        "updated_at": format_datetime(task.updated_at),
    }

File: modules/agent/internal/serializers/task.go
package serializers

import (
	"time"

	"agent/internal/models"
	"github.com/shopspring/decimal"
)

func decimalToString(d *decimal.Decimal) *string {
	if d == nil {
		return nil
	}
	s := d.String()
	return &s
}

// TaskStatusBrief represents status/priority for task
type TaskStatusBrief struct {
	ID   int64  `json:"id"`
	Name string `json:"name"`
}

// TaskEntityInfo represents the linked entity for a task
type TaskEntityInfo struct {
	Type        *string `json:"type"`
	ID          *int64  `json:"id"`
	DisplayName *string `json:"display_name"`
	URL         *string `json:"url"`
	LeadType    *string `json:"lead_type"`
}

// TaskResponse represents a task
type TaskResponse struct {
	ID             int64            `json:"id"`
	Title          string           `json:"title"`
	Description    *string          `json:"description"`
	Status         *TaskStatusBrief `json:"status"`
	Priority       *TaskStatusBrief `json:"priority"`
	Entity         TaskEntityInfo   `json:"entity"`
	StartDate      *string          `json:"start_date"`
	DueDate        *string          `json:"due_date"`
	CompletedAt    *string          `json:"completed_at"`
	EstimatedHours *float64         `json:"estimated_hours"`
	ActualHours    *float64         `json:"actual_hours"`
	Complexity     *int16           `json:"complexity"`
	SortOrder      int              `json:"sort_order"`
	AssignedTo     *IndividualBrief `json:"assigned_to"`
	CreatedAt      string           `json:"created_at"`
	UpdatedAt      string           `json:"updated_at"`
}

func formatDate(t *time.Time) *string {
	if t == nil {
		return nil
	}
	s := t.Format("2006-01-02")
	return &s
}

func formatDateTime(t time.Time) string {
	return t.Format(time.RFC3339)
}

func decimalToFloat(d *decimal.Decimal) *float64 {
	if d == nil {
		return nil
	}
	f, _ := d.Float64()
	return &f
}

// TaskToResponse converts Task model to response
func TaskToResponse(task *models.Task) TaskResponse {
	return TaskToResponseWithEntity(task, nil)
}

// TaskToResponseWithEntity converts Task model to response with entity info
func TaskToResponseWithEntity(task *models.Task, entityInfo *TaskEntityInfo) TaskResponse {
	resp := TaskResponse{
		ID:             task.ID,
		Title:          task.Title,
		Description:    task.Description,
		StartDate:      formatDate(task.StartDate),
		DueDate:        formatDate(task.DueDate),
		CompletedAt:    nil,
		EstimatedHours: decimalToFloat(task.EstimatedHours),
		ActualHours:    decimalToFloat(task.ActualHours),
		Complexity:     task.Complexity,
		SortOrder:      task.SortOrder,
		CreatedAt:      formatDateTime(task.CreatedAt),
		UpdatedAt:      formatDateTime(task.UpdatedAt),
		Entity:         TaskEntityInfo{},
	}

	if entityInfo != nil {
		resp.Entity = *entityInfo
	}

	if task.CompletedAt != nil {
		s := task.CompletedAt.Format(time.RFC3339)
		resp.CompletedAt = &s
	}

	if task.CurrentStatus != nil {
		resp.Status = &TaskStatusBrief{
			ID:   task.CurrentStatus.ID,
			Name: task.CurrentStatus.Name,
		}
	}

	if task.Priority != nil {
		resp.Priority = &TaskStatusBrief{
			ID:   task.Priority.ID,
			Name: task.Priority.Name,
		}
	}

	if task.AssignedTo != nil {
		resp.AssignedTo = &IndividualBrief{
			ID:        task.AssignedTo.ID,
			FirstName: task.AssignedTo.FirstName,
			LastName:  task.AssignedTo.LastName,
			Email:     task.AssignedTo.Email,
		}
	}

	return resp
}
